### PR TITLE
8322846: Running with -Djdk.tracePinnedThreads set can hang

### PR DESCRIPTION
**Pull request by AlanBateman**
-Djdk.tracePinnedThreads is a debugging option that dates from early development in the loom repo to identify pinned threads. It has several issues and this tracing option will eventually be removed (use the JFR events instead). Several hangs have been reported when running with the system property set. The "hangs" stem from the onPinned callback executing while the virtual thread is in a transition state (typically parking). If the virtual parks while printing the stack trace then it works like a nested park where the thread state is never restored. Contention on the System.out can also lead to deadlock when there are platform and pinned virtual threads printing to System.out around the same time.

This PR brings over the changes from the loom repo to avoid these hangs. The changes mean the stack trace is only printed to System.out when the PrintStream lock can be acquired without blocking. It also restores the thread state after printing. An alternative to not printing traces would of course be to queue the traces so they are printed by another thread but this is just adding complexity for a debugging option that we want to go away.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322846](https://bugs.openjdk.org/browse/JDK-8322846): Running with -Djdk.tracePinnedThreads set can hang (**Bug** - P3)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17221/head:pull/17221` \
`$ git checkout pull/17221`

Update a local copy of the PR: \
`$ git checkout pull/17221` \
`$ git pull https://git.openjdk.org/jdk.git pull/17221/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17221`

View PR using the GUI difftool: \
`$ git pr show -t 17221`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17221.diff">https://git.openjdk.org/jdk/pull/17221.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17221#issuecomment-1874167038)

### CODE REVIEW

**Comment by jaikiran** [comment]
Hello Alan, I haven't yet looked in detail the changes, so this merely a comment based on first glance:

> The changes mean the stack trace is only printed to System.out when the PrintStream lock can be acquired without blocking.

I guess that then means that the some pinned threads might not even appear in this generated thread dumps. I can't think of a way where we could at least print something that would at least hint that some pinned threads weren't dumped in the output. This reduces the utility of this system property which is unfortunate, but I do understand the reason.

**Comment by AlanBateman** [comment]
> I guess that then means that the some pinned threads might not even appear in this generated thread dumps.

Just to be clear. The changes here have no impact on thread dumps, the thread dump generated by jcmd Thread.dump_to_file has all virtual threads, including pinned threads.

As regards jdk.tracePinnedThreads. We want this system property to go away, it's just unfortunate that it seems to be widely used.

**Comment by jaikiran** [comment]
Hello Alan,

> > I guess that then means that the some pinned threads might not even appear in this generated thread dumps.
> 
> Just to be clear. The changes here have no impact on thread dumps, the thread dump generated by jcmd Thread.dump_to_file has all virtual threads, including pinned threads.

I didn't word my previous comment correctly. As you say, the `jdk.tracePinnedThreads` system property doesn't play a role in the thread dumps that generated through the jcmd command. This  system property just enables auto generation of a thread dump of a virtual thread when that virtual thread gets pinned. That thread dump logging happens from within that virtual thread, which is what I understand to be the reason behind the potential hang.

My understanding is that the `jcmd Thread.dump_to_file` command, when invoked, does the thread dump generation (for all threads) from within the Attach Listener thread which is a platform thread and thus isn't susceptible to this issue of hanging. 

> 
> As regards jdk.tracePinnedThreads. We want this system property to go away, it's just unfortunate that it seems to be widely used.

I think it was the ease of use (unlike having to start JFR recording) and the "event driven" automated thread dump (it gets logged only when the virtual thread is pinned unlike having to manually generate thread dumps) which made it attractive, especially during the early testing of virtual threads with other parts of the JDK.

**Review by jaikiran — APPROVED** [review]
Overall this change looks OK to me. I'm not too familiar with the (virtual) thread states so I can't say if there's any other implication of changing the virtual thread state to RUNNING when we are printing this thread dump.

I've a trivial review comment about the new test method introduced in this PR, which I have added inline.

File: test/jdk/java/lang/Thread/virtual/TracePinnedThreads.java [hunk 1]
```diff
@@ -78,8 +79,66 @@ void testPinnedCausedBySynchronizedBlock() throws Exception {
...
+        String output = run(C::new);
+        assertContains(output, "reason:NATIVE");
+        assertContains(output, "<clinit>");
+    }
+
+    /**
+     * Test contention writing to System.out when pinned.
+     */
+    @Test
+    void testContention() throws Exception {
```
Comment (jaikiran) [inline test/jdk/java/lang/Thread/virtual/TracePinnedThreads.java:105]: I think this test method  expects to have all the launched threads to print to System.out without hanging. If it hangs then we expect the test run to timeout with a jtreg timeout. Given the context of this PR in which this test method is being introduced, I think I can understand what it does. But maybe we should update the test method's comment to say that this method verifies that contention when writing to System.out when pinned doesn't cause any hangs?

**Review by jaikiran — APPROVED** [review]

**Pull request closed without merging**